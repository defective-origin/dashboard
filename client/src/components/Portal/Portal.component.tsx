import React from 'react'
import { Portal as MuiPortal } from '@mui/base'

// ---| core |---
import { cn } from 'tools'
import { useFunc } from 'hooks'

// ---| pages |---
// ---| screens |---
// ---| components |---

// ---| self |---
import css from './Portal.module.scss'

export type PortalName = 'page-name' | 'page-actions'

export const initPortalKey = (name: PortalName) => `portal-${name}`

export type PortalProps = {
  name: PortalName
  content?: React.ReactNode
  children?: React.ReactNode
}

/**
 * Component description.
 *
 * How to use
 * @example
 * <Portal />
 */
export function Portal(props: PortalProps): JSX.Element {
  const { name, content, children = content } = props
  const initContainer = useFunc(() => document.getElementById(initPortalKey(name)))

  return <MuiPortal container={initContainer}>{children}</MuiPortal>
}

Portal.displayName = 'Portal'

export type PortalContainerProps = {
  name: PortalName
  className?: string
  children?: React.ReactNode
}

export function PortalContainer(props: PortalContainerProps): JSX.Element {
  const { name, children, className, ...otherProps } = props
  const _className = cn(css.PortalContainer, className)

  return <div id={initPortalKey(name)} className={_className} {...otherProps}>{children}</div>
}

PortalContainer.displayName = 'PortalContainer'

Portal.Container = PortalContainer

export default Portal
