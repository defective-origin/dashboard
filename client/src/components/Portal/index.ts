/* INJECT_IMPORT_PLACE */

/* INJECT_DEFAULT_EXPORT_PLACE */
export { default } from './Portal.component'

/* INJECT_EXPORT_PLACE */
export * from './Portal.component'
