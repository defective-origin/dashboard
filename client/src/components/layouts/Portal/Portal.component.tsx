import React from 'react'

// ---| components |---
import Block, { BlockProps } from 'components/layouts/Block'

// ---| core |---
import { cn } from 'tools'
import { PortalReturnOptions, usePortal } from 'hooks'

// ---| self |---
import css from './Portal.module.scss'

export type PortalName = 'page-name' | 'page-extra' | 'page-nav'

export const initPortalKey = (name: PortalName) => `portal-${name}`

export type PortalProps = {
  name: PortalName
  disable?: boolean
  content?: React.ReactNode
  children?: React.ReactNode
}

/**
 * Allows portal content if into portal container.
 *
 * How to use
 * @example
 * <Portal name='page-name'>Content</Portal>
 * <Portal name='page-name' content="Content" />
 */
export function Portal(props: PortalProps): PortalReturnOptions {
  const { name, disable, content, children = content } = props

  return usePortal(children, { ref: () => document.getElementById(initPortalKey(name)), disable })
}

Portal.displayName = 'Portal'

export type PortalContainerProps = BlockProps & {
  name: PortalName
}

/**
 * Create portal container.
 *
 * How to use
 * @example
 * <Portal.Container name='page-name' />
 */
Portal.Container = function PortalContainer(props: PortalContainerProps) {
  const { name, children, className, ...otherProps } = props
  const _className = cn(css.PortalContainer, className)

  return <Block id={initPortalKey(name)} className={_className} {...otherProps}>{children}</Block>
}

export default Portal
