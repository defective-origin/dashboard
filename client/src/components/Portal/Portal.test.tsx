// ---| tests |---
import { render } from 'tests'

// ---| self |---
import Portal from './Portal.component'
// import usePortal from './Portal.hook'
// import PortalProvider, Portal from './Portal.context'

describe('[Portal] component', () => {
  it('should render component', () => {
    const container = render(<Portal />)

    expect(container.snapshot()).toMatchSnapshot()
  })
})

// describe('[Portal] hook', () => {
//   it('should return options', () => {
//     const { result } = renderHook(() => usePortal())
// 
//     expect(result.current).toBe(null)
//   })
// })

// describe('[Portal] context', () => {
//   const wrapper = (props: PortalProps) => <PortalProvider { ...props } />
// 
//   it('should return options', () => {
//     const { result } = renderHook(() => usePortal(), { wrapper })
// 
//     expect(result.current).toBeTruthy()
//   })
// })
