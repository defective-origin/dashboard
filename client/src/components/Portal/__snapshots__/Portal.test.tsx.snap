// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[Portal] component > should render component 1`] = `<DocumentFragment />`;
